### DISCUSSION TITLE
ggplot2 extension packaging best practices workshop requirements

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#53] [Q&A]
what would be the requirements of a useful workshop on this topic?  What's the right level of assumed knowledge?  Who ideally should come (learner side and expert side).

**Comment by teunbrand** [comment]
I'm not sure what you mean with requirements; is it infrastructural like zoom meetings or more abstract like learning goals? (or another option I haven't considered)
I'd think the audience would be from people who've already written their first ggproto classes up to people who have packages on CRAN but are afraid their code is fragile. I suppose once you've collected the actual best practises, your expert could by anyone who can explain *why* these practises are recommended after a bit of research.

**Reply by EvaMaeRey** [reply]
learning goals, people etc.  I put https://github.com/EvaMaeRey/ggtedious together once...  The idea was to build package from scratch and that is probably too time consuming.  I think having a bare-bones package to start from (no test, first-cut examples) is probably the right way forward.  Probably the focus should be on differentiators for gg v. other packages in terms of testing and maintenance.

**Reply by EvaMaeRey** [reply]
I think geom_circleline (like ridgeline but semi transparent circles that sit on a line, I've seen recently w/o extension in the wile) would be a nice target functionality.

**Reply by yjunechoe** [reply]
I feel like you talked about this before, but is there a reason why you prefer the `readme2pkg` approach as opposed to the standard `devtools` approach (as outlined in the [r-pkgs book](https://r-pkgs.org/whole-game.html))? Genuinely curious because I've never used (or even seen!) `readme2pkg` before!

**Reply by EvaMaeRey** [reply]
It's a literate programming approach to package building.  Other implementations are litr and fusen.  The main case for it is that it's narrative-preserving, which could be nice for workshop.  The functions are in sequence and can live in some narrative context.  @cynthiahqy I think has used litr for these reasons. 

All that is done in the case of the readme 2 package approach is moving readme code chunks to .R files in R directory or as needed. And I put devtools/usethis tools commands in readme too (as with litr and fusen, but some with eval = F, cause I think I use more interactively), to kind of remind myself of what needs to happen.

But it is one more, non-standard, thing to be doing.   I don't think it's make or break for a workshop. Maybe it would just feel too distracting.

**Reply by yjunechoe** [reply]
Yeah I think that's fair, but probably a reasonable Q to expect from the audience as well (esp. for the experts).

Speaking of the audience of experts, I'd also be mindful of the upper end of the experience spectrum too. If you get too many "experts" in the room, you might get stuck in an endless debate about whether one "best practice" you're trying to teach them is indeed a best practice at all 😆 (expanding on Teun's point - experts may also have strong feelings about why _their own_ practices should be recommended)

**Reply by EvaMaeRey** [reply]
I want to go to the debates one!  Sounds exciting!  I think folks like Teun or Thomas are the ideal experts and can speak to if  tests are appropriate and package is well structured for vanilla extension (w/ a reverse dependency lens) :-).   

I used to pull code into README.Rmd from R scripts, to talk about it, that could be a post-hoc way to still use the readme as a workshop narrative container - maybe less objectionable.  Or into another document.

**Reply by teunbrand** [reply]
I don't think there really is something as a collection of best practises currently though. In the ggplot2 repo, there isn't really a programming guideline. Most of the stuff Thomas and me would recommend will probably be to not do the stuff that has annoyed us when investigating revdep failures. Which yeah, I guess they are best practises in some sense, but it is still more of a wild grab-bag than a principled approach that you'd like to convey in a workshop. Also most revdep failures are not in extension packages, but in packages that just use ggplot2 in weird ways.

> The functions are in sequence and can live in some narrative context

For reasons I can't quite articulate well, this just doesn't mesh very well with my brain. In my head, putting everything in a linear structure makes everything equally important. For some reason it prompts my brain to want to understand everything all at once, instead of small nuggets of understanding that I can connect in a modular way. I'm not saying it is better or worse than other approaches, just that I personally have a hard time digesting code with it.

**Reply by yjunechoe** [reply]
I think I share this intuition as well, especially re: "makes everything equally important" (warning: opinionated thoughts ahead!). One concrete context is how the current linear narrative in `ggtedious` starts with defining a standalone compute function and builds up to constructing a ggproto out of it. At least for me, when I'm write a ggproto extension, I typically start by subclassing the parent ggproto I want to inherit from using `ggproto()` and "extend" its various methods with `= \(...) browser()` to inspect and identify the appropriate extension point. Once I know which method I want to extend, I start writing the function body _within_ the `ggproto()` call for quick prototyping and iteration. Then, finally, if I'm satisfied and want to take the extra step, I would factor out the method as a standalone function and write unit tests for its behavior.

So while the ability to encapsulate an internal logic like compute as a stand-alone pure function is conceptually very powerful, it's not as important to me so early in the process (at least for my tastes, it feels a bit like premature optimization). The function-first strategy certainly has pedagogical usefulness (it's very empowering to know that "ggproto methods are just data wrangling functions") and your recipes showcase that well. But when it concerns the workflow of developing an extension package in practice, I think there are different priorities involved there. Kinda like how every step to solving the problem is important for you to know when you're just learning something, but once you've distilled the knowledge you can leverage that to move quicker because you've developed heuristics, mental shortcuts, etc.

**Reply by EvaMaeRey** [reply]
> At least for me, when I'm write a ggproto extension, I typically start by subclassing the parent ggproto I want to inherit from using ggproto() and "extend" its various methods with = \(...) browser() to inspect and identify the appropriate extension point. Once I know which method I want to extend, I start writing the function body within the ggproto() call for quick prototyping and iteration. Then, finally, if I'm satisfied and want to take the extra step, I would factor out the method as a standalone function and write unit tests for its behavior.

I think it'd be really instructive to watch your workflow!

**Reply by EvaMaeRey** [reply]
> it prompts my brain to want to understand everything all at once, instead of small nuggets of understanding that I can connect in a modular way. 

Enumeration is meant to help modularize but I guess that's not working.  I think it just might be very hard for an outsider to see how everything fits together without a guide, which is one of the readme as guide motivations.  But, again, not make or break and it's sounding like it'd be a distraction at this point.

**Reply by yjunechoe** [reply]
> I think it just might be very hard for an outsider to see how everything fits together without a guide, which is one of the readme as guide motivations.

I think it might help to more strongly disentangle the "developing extension pkgs" workshop from the "writing ggproto extensions" workshop. I think you might end up stretching yourself too thin with the pkg dev/best practices workshop if you also try to simultaneously introduce people to the workflow of writing a ggproto extension and walking them through that narrative step-by-step.

So I wonder if the workshop would benefit from having a more narrow scope and strictly being geared towards folks who already have an extension in mind (either of their own or from reading your reciples), at which point the question is "how do I polish/organize all this code appropriate for a package?"

**Reply by EvaMaeRey** [reply]
Yeah.  I definitely think just having some working functionality makes sense to get going on a packaging-focused workshop.  Some knowledge is assumed!  And yes, writing ggproto extension workshop as a prestep sounds right.  I think there's a lot of under-explored value in package-free extension, and the whole kit and cabooddle style workshop (from 0 -> functions -> {gg*} -> CRAN-ready) doesn't really send that message.

**Reply by rsh52** [reply]
FWIW I also think that separation would work wonders. CRAN-readiness can definitely be its own thing and the place where things like standard/suggested test development exists.

**Reply by cynthiahqy** [reply]
Super interesting discussion!
**RE: litr/lusen/readmepkg**, I tried litr for documenting my design thinking when I was iterating through different interface ideas (argument names etc), mostly because I often leave things and then come back later having totally forgotten that I've already tried some idea or thought through why some interface choice would be confusing. But it did get clunky at some point to re-render a whole document or set of documents to fix a small bug or test, and so I think it makes sense to migrate from the literate programming structure to a more standard `devtools` approach at some point -- maybe as you move from design/prototyping to publishing/maintaining the package.

**Reply by cynthiahqy** [reply]
I also second the suggestion of a stand-alone "ggproto extensions workshop"! 

I think collecting & synthesising best practices for a) CRAN-ready packages in general, b) designing grammar of graphic extension packages generally and specifically in ggplot2 is a different can of worms to supporting people in a workshop format to write their first ggproto extension (which I definitely want to attend!).

Recently, I've found that I get the most value in workshops from the guided walk-throughs / having synchronous support to get through the initial "not even knowing what to google" stage of learning a new skill. This is even though I am much more interested in discussions about best practices and design principles, and I feel an almost childish petulance about the unreasonable effectiveness of being in a room/zoom session for getting over the initial hump of learning a new skill. 

I'm also increasingly of the opinion that teaching workshops are not the best format for disseminating best practices -- i.e. written guides like [Tidy design principles](https://design.tidyverse.org/) or talks like Thomas's [posit::conf talk](https://youtu.be/mtUKYGvkXm8?feature=shared) are more scalable across people/time. Though I do think discussion panels are a great format for surfacing expert opinions etc. I've also been trying to design a survey study of existing ggplot extensions packages (i.e. retrieving ggplot packages from CRAN/R-Universe/GitHub, and doing some programmatic analysis of code patterns) for documenting "common practices" -- though it's somewhat of a next year project because I have a thesis to submit very soon.

**Reply by EvaMaeRey** [reply]
Interesting points about best practice and where to effectively disseminate.   hmm.

I never used litr or fusen, but have felt that  --- re-render a whole document or set of documents to fix a small bug or test --- pain and rmd fiddly-ness.  I tried ease pain a little by live retrieval of a code chunk, so you can update package elements independently.  Uses the rstudioapi package and litedown. So in rstudio, chunk_to_r("theme_chalkboard") in a live session would send the contents of the chunk 'theme_chalkboard' to a file theme_chalkboard.R in the R folder.  It felt a bit better when that was in place, but still probably the most useful for small projects.  I think it is worth saying that you can totally ignore the readme if you dislike the idea of orienting yourself to the package that way.

Opened a discussion thread about ggproto extensions workshop #55 for folks thoughts on that too.  🙃  

Thomas's talk is really good. 🤯

**Comment by rsh52** [comment]
✋ Learner side! 

**Testing**: Coming from package development for analytics-focused packages, I find it challenging to write a test suite for visualization that promotes not only confidence to users, but mostly to myself that I'm accounting for the wide variety of use cases that come with ggplot2. 

In an analytics package, you would consider writing a sample dataset, feeding it your function, and expecting a defined output. Then expanding on that for variation and edge cases. It's much easier to wrap my head around comparing values/dataframes than outputs of, say, everyone's favorite ggswim package.

Like Teun mentioned, I know I'm afraid of writing fragile code yet outside of `vdiffr` comparisons and under-the-hood data manipulations, it's hard to rigidly define expectations in a test suite. Plus it seems like if you have ideally adhered to the grammar of graphics much of your testing faith should be placed in ggplot2 itself.

**Failing Upwards**: I'm happy to be the poster child for how to take a bunch of wrong turns before making an extension package. Here are some things I tried, picked up, and dropped along the way:

- When is it useful to incorporate a `print` or `build` method? And when should I more importantly _not_ have one?
- Do I _really_ need to make that into a geom?
- What are the core things I need to know when inspecting all those pieces of a layer?
- `grid` seems terrifying, when do I need to go to battle with it?
- Why I should avoid changing layers after they've been created / how I fell down the hole of `ggnewscale`/`ggplot.add` and how this group dug me out.

All this being said, the folks in this discussion are for more advanced than I am and maybe some of these things are better tailored to a more elementary ggplot2 workshop. IMO though, there's an opportunity to target an audience that is versed in package development, fluent enough in the ggplot2 text chapters 1-17, has an idea for an extension, but doesn't know the best practices for making one.

**Reply by teunbrand** [reply]
Well don't hesitate to ask here if you run into something obtuse. Chances are we might have run into similar things or people in the future might want to look into how you solved a thing :)

**Reply by EvaMaeRey** [reply]
In addition to providing so much specificity about what might be useful topics, I think you really make the case for a narrowly scoped discussion/workshop on packaging considerations.   As for myself, I think having more examples and ideas around what constitutes good tests in extension context as well as  gg-style error messages, warnings would be a useful focus.

**Comment by EvaMaeRey** [comment]
I'd kind of forgot about this great discussion 🤩, and really happy to come back to it and refresh on everyone who commented weighed in here.  @cynthiahqy @rsh52 @teunbrand @yjunechoe 
 
I'm back here, after a search of the discussions, because I started working on a 'packaging considerations' demo package: https://github.com/EvaMaeRey/easy.geom.recipes.package   

As kind of separates extension from package in that it assumes that you've gone through the 'easy geom recipes', or similar, and in so doing you'd have familiarity with some fundamental extension moves.   

And, as discussed (debated?) in some detail in this thread, the action does happen in the readme 😬 - code is moved from code chunks to the required folders using [knitrExtra (on github)](https://github.com/EvaMaeRey/knitrExtra).  But it doesn't claim to be the right way to do things - it's for convenience.  I should probably add some disclaimers, e.g. one problem is people that want to contribute to your code could be confused by the set up.  Maybe I should ask the litr and fusen folks about their experiences.  

I wonder @rsh52 if you would be willing to weigh in on self-proclaimed 'learner side', but also someone who's managing a super useful package, regarding tests? I hope to compose some tests soon - totally blank right now.   I think from @yjunechoe, that for Stats you might not need to do a visual test, and could just test the compute?  I feel like it would be nice if that is indeed the case.  After a first cut check with @teunbrand as the authority on good and terrible extension package tests! 🤞😁

**Reply by rsh52** [reply]
Hi! I still am open to learning more about what good unit tests look like for visualization packages. Right now I have a decent smattering testing for layering functionality and low-level manipulations in the ggswim package, but otherwise I rely heavily on `vdiffr` and snapshots to see if there are any egregious changes between expected outputs.

**Reply by teunbrand** [reply]
Wait what am I to check?

**Reply by EvaMaeRey** [reply]
I guess what I'm wondering is if @rsh52 would critique my toy package's test, and then @teunbrand, you critique the critique?  🤷‍♀️

I just wrote this one test so far... 

<img width="757" height="540" alt="image" src="https://github.com/user-attachments/assets/1c4caece-36ad-418b-b64f-af82445e0fc2" />

It's a very narrowly scoped package, just creating user-facing geom_means and stat_means alliases 🙃, but I guess I wonder what other tests you all might suggest?

**Reply by teunbrand** [reply]
I'd typically recommend one round-trip-test where you use vdiffr to snapshot your plot with your custom class.
Then if you have branching logic inside your class, you can test as many branches as you'd like, but you may be able to test the plot data instead of using snapshots.
If your constructor does anything fancy, you probably can test for this by prying the parameters out of the resulting Layer object, or whatever you're constructing, and see if the wrangling worked.

**Reply by EvaMaeRey** [reply]
> round-trip-test  👀

**Reply by teunbrand** [reply]
Maybe round-trip isn't the best name for this. The point is to do the entire song and dance of rendering a plot and checking that it looks like desired. That checks for a bunch of contingencies you might not have tought up yourself and also serves as a 'canary in a coalmine' if changes are coming upstream.

**Reply by rsh52** [reply]
I guess this confirms my suspicions on testing here, the snapshot test is pretty comprehensive for checking any A to B behaviors that go astray. For ggswim I used some simple tests for confirming attributes and classes are passed along correctly and then the snapshot was the last piece of the puzzle: https://github.com/CHOP-CGTInformatics/ggswim/blob/main/tests/testthat/test-geom_swim_lane.R

**Reply by EvaMaeRey** [reply]
I like 'round trip' I wasn't sure if it was an expression that I should already know. It evokes 'do the thing that you plan to do with the package - take the tests all the way to the use case' @teunbrand   @rsh52 thanks for the references.
